providers/flagd/pkg/service/service.go

```diff
@@ -193,66 +380,173 @@ func handleError(err error) openfeature.ResolutionError {
 	return openfeature.NewGeneralResolutionError(err.Error())
 }
 
-// EventStream emits received events on the given channel
-func (s *Service) EventStream(
-	ctx context.Context, eventChan chan<- *schemaV1.EventStreamResponse, maxAttempts int, errChan chan<- error,
-) {
-	delay := s.baseRetryDelay
-	var err error
-	for i := 1; i <= maxAttempts; i++ {
-		s.logger.V(logger.Debug).Info("attempt at connection to event stream", "attempt", i)
-		i, err = s.eventStream(ctx, eventChan, i)
-		if i == 1 {
-			delay = s.baseRetryDelay // reset delay if the connection was successful before failing
-		}
-		if err != nil && i <= maxAttempts {
+func (s *Service) EventChannel() <-chan of.Event {
+	return s.events
+}
+
+// startEventStream - starts listening to flagd event stream with retries.
+// If retrying is exhausted, an event with openfeature.ProviderError is emitted.
+func (s *Service) startEventStream(ctx context.Context) {
+	var delay = s.baseRetryDelay
+	var attempts = 1
+
+	// wraps connection with retry attempts
+	for attempts <= s.maxRetries {
+		s.logger.V(logger.Debug).Info("connecting to event stream")
+		err := s.streamClient(ctx)
```

Comment:
I think here we missed resetting the connection attempts if we successfully connect.

Reply:
This handled at the else block [^1]

And error handling the stream means there's something wrong, which we should retry. This could be a connection loss, packet errors or something else.

However, if there's no error, which can happen with a provider shutdown, stream end, we reset attempts and starts fresh connection establishment attempts 

[^1]: https://github.com/open-feature/go-sdk-contrib/blob/470ee860a64ccade08b6e764962e5b9d50c0a210/providers/flagd/pkg/service/service.go#L397-L409

Reply:
Maybe I am misread, but I think `stream.Receive()` exits only when the stream ends. Hence, if I fail two times to connect, at the third attempt I can connect, after a while the connection drops, the retry starts at 3 instead of 0 🤔

Reply:
Yup, that's something that the current logic doesn't handle. I will improve this

Reply:
Improvement done with this commit - https://github.com/open-feature/go-sdk-contrib/pull/317/commits/44870f92ea07f9923f2ef3f47acf7ca500aad2c4 

Retry counter and sleep delay now live in its own structure